firstvoices/backend/tests/test_async/test_async.py

```diff
@@ -0,0 +1,129 @@
+import json
+from unittest.mock import ANY
+
+import pytest
+
+from backend.models.constants import AppRole, Visibility
+from backend.tests import factories
+from backend.tests.test_apis.base_api_test import BaseApiTest
+
+
+class TestCustomOrderRecalculatePreview(BaseApiTest):
+    API_DETAIL_VIEW = "api:site-detail"
+    SUPERADMIN_PERMISSION = "views.has_custom_order_access"
+
+    @pytest.fixture
+    def site(self):
+        return factories.SiteFactory.create(slug="test", visibility=Visibility.PUBLIC)
+
+    @pytest.fixture
+    def alphabet(self, site):
+        return factories.AlphabetFactory.create(site=site)
+
+    @pytest.mark.django_db
+    def test_recalculate_preview_get_404(self):
+        user = factories.get_app_admin(role=AppRole.SUPERADMIN)
+        self.client.force_authenticate(user=user)
+
+        response = self.client.get(
+            f"{self.get_detail_endpoint('missing-slug')}dictionary-cleanup/preview"
+        )
+        assert response.status_code == 404
+
+    @pytest.mark.django_db
+    def test_recalculate_preview_post_404(self):
+        user = factories.get_app_admin(role=AppRole.SUPERADMIN)
+        self.client.force_authenticate(user=user)
+
+        response = self.client.post(
+            f"{self.get_detail_endpoint('missing-slug')}dictionary-cleanup/preview"
+        )
+        assert response.status_code == 404
+
+    @pytest.mark.django_db
+    def test_recalculate_preview_get_403(self, site):
+        user = factories.get_non_member_user()
+        self.client.force_authenticate(user=user)
+
+        response = self.client.get(
+            f"{self.get_detail_endpoint(site.slug)}dictionary-cleanup/preview"
+        )
+        assert response.status_code == 403
+
+    @pytest.mark.django_db
+    def test_recalculate_preview_post_403(self, site, alphabet):
+        user = factories.get_non_member_user()
+        self.client.force_authenticate(user=user)
+
+        response = self.client.post(
+            f"{self.get_detail_endpoint(site.slug)}dictionary-cleanup/preview"
+        )
+        assert response.status_code == 403
+
+    @pytest.mark.django_db
+    def test_recalculate_preview_get_empty(self, site, alphabet):
+        user = factories.get_app_admin(role=AppRole.SUPERADMIN)
+        self.client.force_authenticate(user=user)
+        assert user.has_perm(self.SUPERADMIN_PERMISSION, site)
+
+        response = self.client.get(
+            f"{self.get_detail_endpoint(site.slug)}dictionary-cleanup/preview"
+        )
+        response_data = json.loads(response.content)
+        assert response.status_code == 404
+        assert response_data == {
+            "message": "No recalculation results found for this site."
+        }
+
+    @pytest.mark.django_db(transaction=True, serialized_rollback=True)
```

Comment:
In order for async tests to run sequentially in our test db, `serialized_rollback` must be enabled. This does slow the test down however.

Reply:
Can we adjust the test so they don't need to run sequentially? Sequential tests usually cause multiple problems down the line

Reply:
Would it be as simple as using a new site for each test and creating the necessary data for the test case? That would make the tests easier to understand as well

Reply:
I'll give it a go, but serialized rollback may be required as after any test with transaction=true the default g2p config will be removed.